### PR TITLE
CXX-1045 collection::count() serializes options::count::hint incorrectly

### PR DESCRIPTION
**Pull request by jrassi**
First included commit is an ABI-breaking/API-breaking small refactor.  Second commit is the bugfix itself.

@acmorrow @xdg, PTAL.

### CODE REVIEW

**Comment by xdg** [comment]
Why is the ABI/API break necessary?  I think you need to elaborate more. Think about the 7th rule of [great commit messages](http://chris.beams.io/posts/git-commit/): _Use the body to explain what and why vs. how_

Refactoring and breaking changes probably deserve some explanation.

**Comment by jrassi** [comment]
Updated with expanded commit message bodies, and rebased against master.

**Comment by xdg** [comment]
I really wish we had a stdx::variant, because we could just replace mongocxx::hint entirely.

Unrelated to this PR, I wonder if there's any point trying to compare strings and documents using the key-name-mangling algorithm.

For v4, I wonder if the member should just become a boxed element, rather than two optionals.  Or if mongocxx::hint should be replaced with a boxed element. Thoughts?

File: src/mongocxx/hint.cpp [hunk 1]
```diff
@@ -27,18 +27,12 @@ hint::hint(bsoncxx::document::view_or_value index) : _index_doc(std::move(index)
 hint::hint(bsoncxx::string::view_or_value index) : _index_string(std::move(index)) {
 }
 
-bsoncxx::document::value hint::to_document() const {
```
Comment (jrassi) [inline src/mongocxx/hint.cpp]: Done, see latest commit eb0ea89.

File: src/mongocxx/test/hint.cpp [hunk 2]
```diff
@@ -52,6 +52,22 @@ TEST_CASE("Hint", "[hint]") {
             auto index_doc = builder::stream::document{} << "a" << 1 << builder::stream::finalize;
             REQUIRE(index_hint != index_doc);
         }
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
```
Comment (jrassi) [inline src/mongocxx/test/hint.cpp]: Alternatively, I could keep the tests and remove the MONGOCXX_DEPRECATED declaration until said ticket is resolved.

Comment (xdg) [inline src/mongocxx/test/hint.cpp]: let's keep the tests and remove MONGOCXX_DEPRECATED and have a ticket for doing that later after we figure out how to suppress deprecation warnings in testing.

Comment (jrassi) [inline src/mongocxx/test/hint.cpp]: Done, see latest commit 54ea3f23e27049f412d0f79f50b6055092bf88f2.

Comment (jrassi) [inline src/mongocxx/test/hint.cpp]: Filed https://jira.mongodb.org/browse/CXX-1056 for suppressing deprecation warnings in testing.

**Pull request merged**